### PR TITLE
Close #261:  Add exponential function

### LINKED ISSUE #261: Add exponential function
Add exponential function in the standard library

### PR DESCRIPTION
**Pull request by dedok1997**
Add a new function in the standard library.

### CODE REVIEW

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package pravda.vm.standard
+
+import pravda.vm.Data.Primitive.{BigInt, Int16, Int32, Int8, Number, Uint16, Uint32, Uint8}
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: I think It can be exchanged with simple `...Primitive._`

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+import pravda.vm.Data.Type
+import pravda.vm.VmError.WrongType
+import pravda.vm.WattCounter.CpuArithmetic
+import pravda.vm._
+
+object ExponentialFunction extends FunctionDefinition {
+  override def id: Long = 0x03L
+
+  override def description: String = "Calculate exponential operation first two items in the stack and push result into"
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: Takes two items from the stack, raises the second number to a power of first number and pushes the result to the stack.

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+package pravda.vm.standard
+
+import pravda.vm.Data.Primitive.{BigInt, Int16, Int32, Int8, Number, Uint16, Uint32, Uint8}
+import pravda.vm.Data.Type
+import pravda.vm.VmError.WrongType
+import pravda.vm.WattCounter.CpuArithmetic
+import pravda.vm._
+
+object ExponentialFunction extends FunctionDefinition {
+  override def id: Long = 0x03L
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: Use `val`, `def` recomputes results on every call.

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala (hunk 2 above)
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: `def` -> `val`

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+  override def description: String = "Calculate exponential operation first two items in the stack and push result into"
+
+  override def args: Seq[(String, Seq[Type])] = Seq(
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: `dev` -> `val`

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala (hunk 3 above)
Comment (vovapolu) [reply vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: Also you don't have to write the explicit type of the field, it it's obvious.

Comment (vovapolu) [reply vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: And `override` keyword

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    "x" -> Seq(Data.Type.Int32),
+    "y" -> Seq(Data.Type.Int32)
+  )
+
+  override val returns = Seq(Data.Type.Number)
+  private val div: Double = math.log10(2.0)
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: Very bad naming here. Probably the best solution here just to take `5` as the factor.

File: vm/src/test/resources/stdlib/scall-exponent.sbox [hunk 6]
```diff
@@ -0,0 +1,11 @@
+---preconditions
+watts-limit: 1000
+---expectations
+watts-spent: 144
+stack
+  uint8(4)
```
Comment (vovapolu) [inline vm/src/test/resources/stdlib/scall-exponent.sbox]: Add one more test with overflow. All in all, great tests :+1:

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+  override def apply(memory: Memory, wattCounter: WattCounter): Unit = {
+    implicit class DoubleExp(x: Double) {
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: Overkill. You can define several overloaded functions instead of implicit classes. Implicit classes have some overhead and should be used with care.

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+    implicit class ByteExp(x: Byte) {
+      def **(y: Double): Int = math.pow(x.toDouble, y).toInt
+      def **(y: Byte): Int = math.pow(x.toDouble, y.toDouble).toInt
+      def **(y: Int): Int = math.pow(x.toDouble, y.toDouble).toInt
+      def **(y: Long): Int = math.pow(x.toDouble, y.toDouble).toInt
+      def **(y: Short): Int = math.pow(x.toDouble, y.toDouble).toInt
+    }
+
+    def getCost(x: Double, y: Double): Long = {
+      (math.log10(math.abs(x) + 1) / div * math.abs(y) / 8).toLong + 1
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: Simplify the formula, I understand that you want to take `log_256(x^y)`, but it don't have to be exact.

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    def getBigIntCost(x: scala.BigInt, y: Int): Long = {
+      (x.abs.toString().length / div * math.abs(y) / 8).toLong + 1
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: `if (true.toString.length == 4) { ... }`

Take a look at `.bitCount`, `.bitLength` methods. I'd also simplify this method in terms of constants.

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    }
+    wattCounter.cpuUsage(CpuArithmetic)
+    val a = memory.pop()
+    val b = memory.pop()
+
+    wattCounter.cpuUsage(CpuArithmetic)
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: Why two times?

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+            wattCounter.memoryUsage(getCost(lhs.toDouble, rhs.toDouble))
+            Int32(lhs ** rhs)
+          case Uint16(rhs) =>
+            wattCounter.memoryUsage(getCost(lhs.toDouble, rhs.toDouble))
+            Int32(lhs ** rhs)
+          case Uint32(rhs) =>
+            wattCounter.memoryUsage(getCost(lhs.toDouble, rhs.toDouble))
+            Int32(lhs ** rhs)
+          case BigInt(rhs) =>
+            wattCounter.memoryUsage(getCost(lhs.toDouble, rhs.doubleValue()))
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: It's dangerous to convert `BigInt` to `Double`. It can be evaluated to `POSITIVE_INFINITY`. Write a method that takes two `BigInt`s and then convert the other argument to `BigInt` here.

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala (hunk 8 above)
Comment (vovapolu) [reply vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: Also that's wrong. If we produce `Int32` for example, we don't care about its length. We use 4 bytes in any case.

File: vm/src/test/resources/stdlib/scall-exponent.sbox (hunk 6 above)
Comment (vovapolu) [reply vm/src/test/resources/stdlib/scall-exponent.sbox]: And with double overflow, it can become `POSITIVE_INFINITY `. Interesting to see this :)

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala (hunk 2 above)
Comment (dedok1997) [reply vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: It's more correctly. Thanks

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala (hunk 7 above)
Comment (dedok1997) [reply vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: I think, here we can make our implicit classes as value classes.

Comment (dedok1997) [reply vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: And make it is top-level classes ofc

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+    "x" -> Seq(Data.Type.Int32),
+    "y" -> Seq(Data.Type.Int32)
+  )
+
+  val returns = Seq(Data.Type.Number)
+
+  def apply(memory: Memory, wattCounter: WattCounter): Unit = {
+    import PowerfulImplicits._
+
+    def receivePayment(primitive: Data.Primitive): Data.Primitive = {
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: calculateWatts

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+          case Int32(rhs) =>
+            receivePayment(Number(lhs ** rhs))
+          case Uint8(rhs) =>
+            receivePayment(Number(lhs ** rhs))
+          case Uint16(rhs) =>
+            receivePayment(Number(lhs ** rhs))
+          case Uint32(rhs) =>
+            receivePayment(Number(lhs ** rhs))
+          case BigInt(rhs) =>
+            receivePayment(Number(lhs ** rhs.longValue()))
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: Write tests where you raise small primitive number to the power of bigint.

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+
+  def apply(memory: Memory, wattCounter: WattCounter): Unit = {
+    import PowerfulImplicits._
+
+    def receivePayment(primitive: Data.Primitive): Data.Primitive = {
+      wattCounter.memoryUsage(primitive.volume.toLong)
+      primitive
+    }
+
+    def getBigIntCost(x: scala.BigInt, y: Int): Long = { //FIXME overflow
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: calculateWattsForBigInt

File: vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+    "Takes two items from the stack, raises the second number to a power of first number and pushes the result to the stack."
+
+  val args: Seq[(String, Seq[Type])] = Seq(
+    "x" -> Seq(Data.Type.Int32),
+    "y" -> Seq(Data.Type.Int32)
+  )
+
+  val returns = Seq(Data.Type.Number)
+
+  //TODO ask about overloading
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/standard/ExponentialFunction.scala]: Remove TODO

**Comment by vovapolu** [comment]
Great job! 👍
